GpsDataCapturer/sharedlibrary/src/main/java/com/google/sharedlibrary/gpxfile/GpxWriteHandler.java

```diff
@@ -0,0 +1,126 @@
+                Log.d(TAG, "Writing new file header.");
+                bufferedWriter.write(createFileHeader(Utils.getFormattedCurrentTime(context)));
+            }
+
+            //write the captured gps data to file
+            String trackPointXml = getTrackPointXml(location,formattedTime);
+            bufferedWriter.write(trackPointXml);
+            bufferedWriter.flush();
+            bufferedWriter.close();
+            fileWriter.close();
+            Log.d(TAG, trackPointXml);
+        } catch (Exception e) {
+            Log.e(TAG, "GpxFileWriter.writeGpsData", e);
+        }
+    }
+
+    /**
+     * Create the xml header with version, creator and metadata
+     *
+     * @param formattedStartTime time of on location changed in format
+     * @return A header string
+     */
+    public static String createFileHeader(String formattedStartTime) {
+        StringBuilder header = new StringBuilder();
+
+        header.append("<?xml version='1.0' encoding='UTF-8' ?>");
+        header.append("<gpx version=\"1.1\" creator=\"GpsDataCapturer " + BuildConfig.VERSION_CODE
+                + "\" ");
+        header.append("xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" ");
+        header.append("xmlns=\"http://www.topografix.com/GPX/1/1\" ");
+        header.append("xsi:schemaLocation=\"http://www.topografix.com/GPX/1/1 ");
+        header.append("http://www.topografix.com/GPX/1/1/gpx.xsd\">\n");
+        header.append("<metadata><time>").append(formattedStartTime).append("</time>");
+        header.append("<device>").append(Build.DEVICE).append("</device>");
+        header.append("<id>").append(Build.ID).append("</id>");
+        header.append("<manufacturer>").append(Build.MANUFACTURER).append("</manufacturer>");
+        header.append("<model>").append(Build.MODEL).append("</model></metadata>\n");
+        header.append("<trk>\n");
+        header.append("<trkseg>\n");
+        return header.toString();
+    }
+
+    /**
+     * Generate the xml track point of the location
+     *
+     * @param location      the location captured by GPS
+     * @param formattedTime time of on location changed in format
+     * @return a string of xml track point
+     */
+    private String getTrackPointXml(Location location, String formattedTime) {
+        StringBuilder trackPoint = new StringBuilder();
+
+        trackPoint.append("<trkpt lat=\"")
+                .append(location.getLatitude())
+                .append("\" lon=\"")
+                .append(location.getLongitude())
+                .append("\">");
+
+        if (location.hasAltitude()) {
+            trackPoint.append("<ele>").append(location.getAltitude()).append("</ele>");
+        }
+
+        trackPoint.append("<time>").append(formattedTime).append("</time>");
+
+        trackPoint.append("<speed>").append(
+                location.hasSpeed() ? location.getSpeed() : "0.0").append("</speed>");
+
+        trackPoint.append("<accuracy>").append(
```

Comment:
please add a check for location.hasAccuracy

Reply:
Already there.